---
title: "Picking CCLE cell lines for pathway analysis"
output: github_document
---

This is for an R01 application. We want to study cellular signaling using proteomics. The idea is that current knowledge about signaling is very “fuzzy” due to overlapping of signaling components between pathways, e.g. both EGFR and FGFR signal through the MAPK signaling cascade.  

We were criticized about how model cell lines are picked for the study, which was somewhat arbitrary. So I want to see if I can suggest a way to choose the cells based on the expression level of the pathway members. 

RNA-seq data were downloaded from the CCLE. The data are read counts, so I will have to normalize them. I will use the DESeq2 package to do so.

```{r packages, message=FALSE}
library(tidyverse)
library(ggthemes)
library(DESeq2)
library(robustbase)
```

```{r import CCLE data}
# import the raw count data, there are two rows of unwanted materials, so I will have to skip it
raw.counts <- read.delim("../../CCLE_DepMap_18Q2_RNAseq_reads_20180502.gct",
                         header = TRUE, stringsAsFactors = FALSE,
                         skip = 2)
raw.counts[1:5, 1:5]
```

DESeq2 requires a matrix with gene name as row names, and a separated matrix for annotation containing the column names (which is the cell lines in this case).
```{r data import and normalization using DESeq2}
# make the coutn data matrix
count.data <- as.matrix(raw.counts[, 3:1078])
row.names(count.data) <- raw.counts$Name         # the row names are the Ensembl IDs
count.data[1:5, 1:5]
# annotation matrix, this contain a single column named condition, which is the name of the cell lines
coldata <- as.matrix(data.frame(condition = names(raw.counts[, 3:1078]))) 
row.names(coldata) <- names(raw.counts[, 3:1078])   # also make the cell lines as the col.names
head(coldata, 3)
# create the DESeq object
dds <- DESeqDataSetFromMatrix(countData = count.data,
                              colData = coldata,
                              design = ~condition) # design is just the condition in this case

# then create the normalization factor using estimateSizeFactors()
dds <- estimateSizeFactors(dds)

# retrieve the normalized count using counts(, normalized = TRUE)
nor.CCLE <- raw.counts[, 1:2]
colnames(nor.CCLE) <- c("Ensembl", "Gene.names")
nor.CCLE <- cbind(nor.CCLE, data.frame(counts(dds, normalized = TRUE), stringsAsFactors = FALSE), row.names = NULL)

nor.CCLE[1:5,1:5]
```

```{r import the genes involved regulations according to Reactome}
### signal transduction -> by RTK -> EGFR ###
egfr <- read.delim("../../signaling_genes/by_EGFR.tsv", header = TRUE, stringsAsFactors = FALSE)
egfr <- egfr %>% separate(MoleculeName, into = c("Molecule" , "Gene.name"), sep = " ") %>%
  filter(MoleculeType == "Proteins")
egfr <- egfr$Gene.name[!duplicated(egfr$Gene.name)]

```

```{r imported the normalized gene expression saved in earlier part of this R script}
nor.CCLE <- read.delim("Normalized_CCLE_RNAseq.txt", header = TRUE, stringsAsFactors = FALSE)
```

To see what is the best way of picking the cell lines to study, I want to try Rank sum of the expression, mean expresssion, median and the sum of expression of all genes involved in a particular pathway / stimulation. 
```{r try ranking the pathway components}

##################################################################
##################### the following parts work with EGFR 
##################################################################

### let's try using rank to see whether it is a good way to differentiate the cell lines ###

# copy the df 1st
rank.expression <- nor.CCLE

# then rank order all of the gene expressions, producing a matrix
rank.expression <- apply(rank.expression[, 3:1078], 2, 
                         FUN = rank, ties.method = "min") # rank will order ascendingly, 
                                                          # so lower expression get a smaller rank number
                                                          # ties.method = "min" put the ties to a smaller number

# add the Ensembl IDs and gene names back to the matrix, produce a df
rank.expression <- cbind(nor.CCLE[, 1:2], rank.expression)

# then select genes that are only in the egfr pathway
egfr.rank <- rank.expression[rank.expression$Gene.names %in% egfr, ]

# add up the rank of the egfr pathway components 
egfr.matrix <- data.frame(Ensembl = rank.expression$Ensembl, stringsAsFactors = FALSE)
egfr.matrix <- data.frame(Rank.sum = colSums(egfr.rank[, 3:1078]), stringsAsFactors = FALSE)

```


```{r making the medians, means and sums of expression of the EGFR components}

# make a df containig the normalized expression of egfr components
egfr.ccle <- nor.CCLE[nor.CCLE$Gene.names %in% egfr, ]

#####################
### median ##########
#####################

# calculate the medians and add to the egfr.matrix
egfr.matrix$Median <- robustbase::colMedians(as.matrix(egfr.ccle[, 3:1078]))

#####################
### mean ##########
#####################

egfr.matrix$Mean <- colMeans(as.matrix(egfr.ccle[, 3:1078]))

#####################
### median ##########
#####################

egfr.matrix$Sum <- colSums(as.matrix(egfr.ccle[, 3:1078]))

```

```{r clean up and visualize}

egfr.matrix$Cell.lines <- row.names(egfr.matrix)
egfr.matrix.long <- egfr.matrix %>% gather(key = "Methods", value = "Expressions", 1:4)

ggplot(egfr.matrix.long) +
  geom_point(mapping = aes(x = Cell.lines, y = Expressions)) +
  facet_wrap(~ Methods, scales = "free_y")

# boxplot
ggplot(egfr.matrix.long) +
  geom_boxplot(mapping = aes(x = Methods, y = Expressions))

# get the outliner value, this is the max value of the whisker
max.whisk <- quantile(egfr.matrix$Sum, 0.75) + 1.5 * IQR(egfr.matrix$Sum)

# get the cell lines in the outlying region
Cells <- data.frame(Cell.lines = egfr.matrix$Cell.lines[egfr.matrix$Sum >= max.whisk])
Cells$Signal <- "EGFR"
```

